packages/twenty-server/src/engine/twenty-orm/factories/scoped-workspace-datasource.factory.ts

```diff
@@ -0,0 +1,36 @@
+import { Inject, Injectable, Scope } from '@nestjs/common';
+import { REQUEST } from '@nestjs/core';
+
+import { EntitySchema } from 'typeorm';
+
+import { DataSourceService } from 'src/engine/metadata-modules/data-source/data-source.service';
+import { EnvironmentService } from 'src/engine/integrations/environment/environment.service';
+import { Workspace } from 'src/engine/core-modules/workspace/workspace.entity';
+import { WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
+import { WorkspaceDatasourceFactory } from 'src/engine/twenty-orm/factories/workspace-datasource.factory';
+
+@Injectable({ scope: Scope.REQUEST })
+export class ScopedWorkspaceDatasourceFactory {
+  constructor(
+    @Inject(REQUEST) private readonly request: Request,
+    private readonly dataSourceService: DataSourceService,
+    private readonly environmentService: EnvironmentService,
+    private readonly workspaceDataSourceFactory: WorkspaceDatasourceFactory,
+  ) {}
+
+  public async create(entities: EntitySchema[]) {
+    const workspace: Workspace | undefined = this.request['req']['workspace'];
+
```

Comment:
I think this one is a bit dangerous since we don't provide the schema name if I understood correctly. I guess ScopedWorkspaceDatasourceFactory should throw if the request object is not properly set and the token is not valid.
For other cases where we can't use the datasource without requests, the exception should tell us that we are probably not using the right class (for example if I'm using this in a command) as I don't think we want to use ScopedWorkspaceDatasourceFactory but explicitly the one that takes workspaceId as parameter in those cases (workspaceDataSourceFactory/TwentyORMManager). Wdyt?

Reply:
@magrinj I'd rather not return anything if possible? this datasource won't be usable without a proper schema defined anyway and you would have to use  TwentyORMManager.getRepositoryForWorkspace right?

Reply:
It's a bit complicated on my opinion to return something `null` or `undefined` or throw.
As the datasource is created when we inject it inside the class and we can have authenticated or unauthenticated resolver inside this class.

Reply:
I've tested multiple things, and the "best" things I can do is returning `null` when workspace is not defined, so `@InjectWorkspaceDataSource` or `@InjectWorkspaceReposiroty` can now return `null` value.
We need to be careful with that as decorators doesn't enforce the type of the value under them.